mkosi/vmspawn.py

```diff
@@ -0,0 +1,227 @@
+import contextlib
+import os
+import shutil
+import subprocess
+import sys
+from collections.abc import Mapping
+from pathlib import Path
+
+from mkosi.architecture import Architecture
+from mkosi.config import (
+    ConfigFeature,
+    MkosiArgs,
+    MkosiConfig,
+    OutputFormat,
+    QemuFirmware,
+)
+from mkosi.log import die
+from mkosi.partition import finalize_root, find_partitions
+from mkosi.qemu import (
+    KernelType,
+    QemuDeviceNode,
+    copy_ephemeral,
+    machine_cid,
+    start_swtpm,
+    start_virtiofsd,
+    vsock_notify_handler,
+)
+from mkosi.run import run, spawn
+from mkosi.types import PathString
+from mkosi.util import INVOKING_USER
+
+
+def config_feature_to_tristate(cf: ConfigFeature) -> str:
+    if cf == ConfigFeature.enabled:
+        return "yes"
+    if cf == ConfigFeature.disabled:
+        return "no"
+    return ""
+
+
+def run_vmspawn(args: MkosiArgs, config: MkosiConfig, qemu_device_fds: Mapping[QemuDeviceNode, int]) -> None:
+    if config.output_format != OutputFormat.disk:
+        die(f"{config.output_format} images cannot be booted in systemd-vmspawn")
+
+    if (config.runtime_trees and config.qemu_firmware == QemuFirmware.bios):
+        die("RuntimeTrees= cannot be used when booting in BIOS firmware")
```

Comment:
This seems out of place here? We should fail when `QemuFirmware=bios` I guess

Reply:
Fixed in 594d4fb.